### PR TITLE
BaseClientSession

### PR DESCRIPTION
**Pull request by orlichter1**
## Status
Ready

## Description
Needed a subclass of BaseClient in order to keep a session open so added it to CSP.

## Must have
- [x] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

### CODE REVIEW

**Comment by orlichter1** [comment]
@jochman FYI

File: Scripts/CommonServerPython/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
 ## [Unreleased]
-
+- Added ***BaseClientSession*** - a child of BaseClient class that uses session in order to maintain an open session 
+with the server.
```
Comment (yaakovi) [inline Scripts/CommonServerPython/CHANGELOG.md]: don't break it, this should be a single line.

File: Scripts/CommonServerPython/CommonServerPython.py [hunk 2]
```diff
@@ -2380,7 +2380,138 @@ def _is_status_code_valid(self, response, ok_codes=None):
             if status_codes:
                 return response.status_code in status_codes
             return response.ok
+    class BaseClientSession(BaseClient):
```
Comment (yaakovi) [inline Scripts/CommonServerPython/CommonServerPython.py]: add blank line before the class

File: Scripts/CommonServerPython/CommonServerPython.py [hunk 3]
```diff
@@ -2380,7 +2380,138 @@ def _is_status_code_valid(self, response, ok_codes=None):
...
+        def __init__(self, base_url, *args, **kwargs):
```
Comment (yaakovi) [inline Scripts/CommonServerPython/CommonServerPython.py]: add docstring to the class.

File: Scripts/CommonServerPython/CommonServerPython.py (hunk 2 above)
Comment (yaakovi) [inline Scripts/CommonServerPython/CommonServerPython.py]: should be called "BaseSessionClient"

Comment (jochman) [inline Scripts/CommonServerPython/CommonServerPython.py]: change to BaseSessionClient

File: Scripts/CommonServerPython/CommonServerPython.py [hunk 4]
```diff
@@ -2380,7 +2380,138 @@ def _is_status_code_valid(self, response, ok_codes=None):
...
+            self._session = requests.Session()
+            super().__init__(base_url, *args, **kwargs)
+
+        def _http_request(self, method, url_suffix, full_url=None, headers=None,
```
Comment (jochman) [inline Scripts/CommonServerPython/CommonServerPython.py]: add unit tests (you can roughly copy it from the BaseClient tests.

**Comment by orlichter1** [comment]
@jochman Added. What is that? What did I add?

**Comment by jochman** [comment]
@orlichter1 it's preventing the auto-generation of the documentation from parsing the class as function.

**Comment by orlichter1** [comment]
@jochman How do I know if the build is ok? (you said it always fails)

File: Scripts/CommonServerPython/CommonServerPython.py [hunk 5]
```diff
@@ -2381,6 +2381,166 @@ def _is_status_code_valid(self, response, ok_codes=None):
                 return response.status_code in status_codes
             return response.ok
 
+    class SessionBaseClient(BaseClient):
```
Comment (anara123) [inline Scripts/CommonServerPython/CommonServerPython.py]: why do we need 2 classes? BaseClient and SessionBaseClient?
Cant SessionBaseClient totally replace BaseClient?
Are there cases where SessionBaseClient will behave differently?

**Comment by kirbles19** [comment]
@orlichter1 Done with the changelog. Please check for technical accuracy, and let's discuss.

File: Scripts/CommonServerPython/CommonServerPython.py (hunk 5 above)
Comment (orlichter1) [reply Scripts/CommonServerPython/CommonServerPython.py]: @jochman What do you say? Would BaseClient act the same if we just add session to it?

Comment (jochman) [reply Scripts/CommonServerPython/CommonServerPython.py]: After checking it out, changing the `BaseClient` to a session will not break the unit tests.
@orlichter1 guided to open a new branch and change the `requests.request` to `session`. Then we will check if it breaks tests.

**Pull request closed without merging**